### PR TITLE
[Yul] Format string literals back to quoted string literals

### LINKED ISSUE #7505: Nicely format strings in Yul IR generation
When generating the Yul intermediate code of Solidity programs, strings are always translated to hex numbers. This prevents readability of IR code. Because of that, the code generator should detect if a string can be represented in printable ascii chars (for safety) and use that instead.

## How to implement:

Add a function `std::string formatAsStringOrNumber(std::string const& _value)` in `libdevcore/CommonData.h` that checks if all characters are safe and outputs a quoted string. If there is an unsafe charater, use `"0x" + h256(_value, h256::AlignLeft).hex();` as it has been used before (you need to include `FixedHash.h` for that).

Now use that function in ABIFunctions.cpp:978 (`ABIFunctions::abiEncodingFunctionStringLiteral`) and :993. The same change can be applied to `YulUtilFunctions::conversionFunctionSpecial`.

## How to test:

Run `solc --ir` on e.g.
```
contract c {
function f() external pure returns (string memory) { return "abcabc"; }
}
```

It should not contain `0x6162636162630000000000000000000000000000000000000000000000000000` anymore but `"abcabc"` instead.

## How to add an automated test

Create a copy of the directory `test/cmdlineTests/standard_ir_requested` such that `input.json` contains code similar to the one above. Run `./test/cmdlineTests.sh` to update the expectation file.

You should additionally create a test inside `test/libsolidity/semanticTests` that ends in

```
// ====
// compileViaYul: also
// ----
// f()
```

This test should check that the contract still works as expected. Please also try string literals longer than 32 bytes.

obsoletes https://github.com/ethereum/solidity/issues/7495

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/7505

Finished version of https://github.com/ethereum/solidity/pull/7522 since I couldn't push to the fork.

Depends on https://github.com/ethereum/solidity/pull/7786

### CODE REVIEW

File: libdevcore/CommonData.h [hunk 1]
```diff
@@ -357,6 +356,10 @@ std::string getChecksummedAddress(std::string const& _addr);
 bool isValidHex(std::string const& _string);
 bool isValidDecimal(std::string const& _string);
 
+// @returns a quoted string if all characters are printable ASCII chars,
```
Comment (chriseth) [inline libdevcore/CommonData.h]:
```suggestion
/// @returns a quoted string if all characters are printable ASCII chars,
```

File: libdevcore/CommonData.cpp [hunk 2]
```diff
@@ -155,3 +158,14 @@ bool dev::isValidDecimal(string const& _string)
 		return false;
 	return true;
 }
+
+// Returns a quoted string if all characters are printable ASCII chars,
+// or its hex representation otherwise.
+std::string dev::formatAsStringOrNumber(std::string const& _value)
+{
+	for (auto const& c: _value)
+		if (!isprint(c))
```
Comment (chriseth) [inline libdevcore/CommonData.cpp]: This may be locale-dependent...

**Review by chriseth — CHANGES_REQUESTED** [review]
Please also add a changelog entry.

File: libdevcore/CommonData.cpp (hunk 2 above)
Comment (leonardoalt) [reply libdevcore/CommonData.cpp]: I'm not sure, its docs state that
` In C++, a locale-specific template version of this function (isprint) exists in header <locale>.`

Comment (leonardoalt) [reply libdevcore/CommonData.cpp]: ah, that doesn't mean the negation of your statement

Comment (leonardoalt) [reply libdevcore/CommonData.cpp]: What's the rule for it to be safely printable?

**Comment by leonardoalt** [comment]
The cli test failure is `0xffffffff` vs `0xFFFFFFFF`, not sure why that is..

File: libdevcore/CommonData.cpp (hunk 2 above)
Comment (chriseth) [reply libdevcore/CommonData.cpp]: Oh we should also exclude `"` of course (or escape it).

Maybe go with
```
     digits (0123456789)
    uppercase letters (ABCDEFGHIJKLMNOPQRSTUVWXYZ)
    lowercase letters (abcdefghijklmnopqrstuvwxyz)
    punctuation characters excluding double quote (!#$%&'()*+,-./:;<=>?@[\]^_`{|}~)
    space ( ) 
```

File: test/cmdlineTests/yul_string_format_ascii/output.json [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                switch selector
+
+                case 0x26121ff0
+                {
+                    // f()
+                    if callvalue() { revert(0, 0) }
+                    abi_decode_tuple_(4, calldatasize())
+                    let ret_0 :=  fun_f_9()
+                    let memPos := allocateMemory(0)
+                    let memEnd := abi_encode_tuple_t_string_memory_ptr__to_t_string_memory_ptr__fromStack(memPos ,  ret_0)
```
Comment (axic) [inline test/cmdlineTests/yul_string_format_ascii/output.json:64]: Why is there that weird spacing around the arguments?

Comment (leonardoalt) [reply test/cmdlineTests/yul_string_format_ascii/output.json:64]: Can you check locally if that happens as well?

File: libdevcore/CommonData.cpp (hunk 2 above)
Comment (axic) [reply libdevcore/CommonData.cpp]: Perhaps we are actually safe if `locale` is not included: http://www.cplusplus.com/reference/cctype/isprint/

> For the standard ASCII character set (used by the "C" locale), printing characters are all with an ASCII code greater than 0x1f (US), except 0x7f (DEL).
>
>  In C++, a locale-specific template version of this function (isprint) exists in header <locale>.

**Comment by leonardoalt** [comment]
Updated

File: libdevcore/CommonData.cpp [hunk 4]
```diff
@@ -155,3 +156,14 @@ bool dev::isValidDecimal(string const& _string)
 		return false;
 	return true;
 }
+
+// Returns a quoted string if all characters are printable ASCII chars,
+// or its hex representation otherwise.
+std::string dev::formatAsStringOrNumber(std::string const& _value)
+{
+	for (auto const& c: _value)
+		if (c < 0x1f || c > 0x7e || c == '"')
```
Comment (chriseth) [inline libdevcore/CommonData.cpp]:
```suggestion
		if (c <= 0x1f || c > 0x7e || c == '"')
```

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Fixed case problem.

**Comment by chriseth** [comment]
I have the suspicion, that "nouppercase" does not actually mean "lowercase" as suggested by the name.

My proposal would be to re-implement `toHex` not using iostream and then using it as much as possible in other locations.

**Comment by leonardoalt** [comment]
Depends on https://github.com/ethereum/solidity/pull/7786

**Review by leonardoalt — COMMENTED** [review]
Approved. (I can't click approve)

**Review by chriseth — APPROVED** [review]

File: libdevcore/CommonData.cpp [hunk 5]
```diff
@@ -181,3 +182,15 @@ bool dev::isValidDecimal(string const& _string)
 		return false;
 	return true;
 }
+
+// Returns a quoted string if all characters are printable ASCII chars,
+// or its hex representation otherwise.
+std::string dev::formatAsStringOrNumber(std::string const& _value)
+{
+	if (_value.length() <= 32)
```
Comment (axic) [inline libdevcore/CommonData.cpp:190]: So if it is not printable as text, but is longer than 32 bytes, it is still printed as unprintable text?

File: test/libsolidity/semanticTests/viaYul/string_format.sol [hunk 6]
```diff
@@ -0,0 +1,13 @@
+contract C {
+	function f1() external pure returns (string memory) { return "abcabc"; }
+	function f2() external pure returns (string memory) { return "abcabc`~12345677890- _=+!@#$%^&*()[{]}|;:',<.>?"; }
```
Comment (axic) [inline test/libsolidity/semanticTests/viaYul/string_format.sol:3]: Is this longer than 32 bytes?

Comment (chriseth) [reply test/libsolidity/semanticTests/viaYul/string_format.sol:3]: yes

File: libdevcore/CommonData.cpp (hunk 5 above)
Comment (chriseth) [reply libdevcore/CommonData.cpp:190]: I'll change it to an assertion. String literals can be at most 32 bytes in yul.

**Pull request merged**